File: typescript/src/lexical-terms.ts
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { StorageError } from '@strands-agents/sdk'

import { Buffer } from 'node:buffer'

/** Version of the tokenizer and key encoding (`lexical-v1`), recorded on every manifest. */
export const LEXICAL_TOKENIZER_VERSION = 'lexical-v1'

const WORD_CHARACTER = /^[\p{L}\p{M}\p{N}]$/u
const JOINERS = new Set(['-', '_'])
const JOINER = /[-_]/
const JOINER_RUNS = /[-_]+/
const EDGE_JOINERS = /^[-_]+|[-_]+$/g
const ASCII_UPPERCASE_RUNS = /[A-Z]+/g
const CONTROL_CHARACTER = /\p{Cc}/u
const EDGE_SEPARATOR = /^\p{Z}|\p{Z}$/u
const LONE_SURROGATE = /\p{Surrogate}/u

const TERM_POSTING_TAG = 't|'
const IDENTIFIER_POSTING_TAG = 'i|'
const MANIFEST_TAG = 'm|'

const utf8 = new TextEncoder()

/**
 * Splits `text` into searchable terms (`lexical-v1`).
 *
 * The text is NFC-normalized and split by code point into runs of letters, marks, numbers and the
 * ASCII joiners `-` and `_`; everything else separates runs. Each run loses its leading and trailing
 * joiners and has ASCII `A`–`Z` lowercased (non-ASCII case is kept because JavaScript and Python fold
 * it differently). A run is emitted as a compound term followed, when it contains joiners, by its
 * joiner-separated parts. Terms are de-duplicated keeping the first occurrence.
 *
 * @throws {@link StorageError} if the text is not well-formed Unicode (it contains a lone surrogate)
 */
export function textTerms(text: string): string[] {
  assertWellFormed(text, 'Text')
  const terms = new Set<string>()
  for (const run of wordRuns(text.normalize('NFC'))) {
    for (const term of runTerms(run)) terms.add(term)
  }
  return [...terms]
}

/**
 * Validates an exact-match identifier and returns its NFC form. Identifiers are otherwise kept as
 * given: case-sensitive, with punctuation, leading zeroes and internal spaces preserved.
 *
 * @throws {@link StorageError} if the identifier is not well-formed Unicode, is empty, contains a control
 *   character, or starts or ends with a separator (whitespace) character
 */
export function normalizeIdentifier(identifier: string): string {
  assertWellFormed(identifier, 'Identifier')
  const normalized = identifier.normalize('NFC')
  if (normalized === '') throw new StorageError('Identifier must not be empty')
  if (CONTROL_CHARACTER.test(normalized)) throw new StorageError('Identifier must not contain control characters')
  if (EDGE_SEPARATOR.test(normalized)) {
    throw new StorageError('Identifier must not start or end with a whitespace or separator character')
  }
  return normalized
}

/**
 * Normalizes each identifier and de-duplicates the result, keeping the first occurrence.
 *
 * @throws {@link StorageError} if any identifier is invalid (see {@link normalizeIdentifier})
 */
export function normalizeIdentifiers(identifiers: readonly string[]): string[] {
  return [...new Set(identifiers.map((identifier) => normalizeIdentifier(identifier)))]
}

/** Number of bytes in the UTF-8 encoding of `value`. */
export function utf8ByteLength(value: string): number {
  return utf8.encode(value).byteLength
}

/**
 * Orders strings by their UTF-8 bytes, which is the order DynamoDB sorts string keys in. The default
 * JavaScript comparison orders UTF-16 code units and disagrees for characters outside the BMP.
 */
export function compareUtf8(left: string, right: string): number {
  return Buffer.compare(utf8.encode(left), utf8.encode(right))
}

/** Length-prefixed key segment `"<utf8 bytes>:<value>"`, so adjacent segments can never be confused. */
export function keySegment(value: string): string {
  return `${utf8ByteLength(value)}:${value}`
}

/**
 * Scope tag stored as `lxscope` on indexed base items. Length-prefixed rather than raw so the root
 * scope (`''`) never becomes an empty-string expression value.
 */
export function scopeSegment(scope: string): string {
  return keySegment(scope)
}

/** Index partition key of the postings of `term` within `scope`. */
export function termPostingPk(scope: string, term: string): string {
  return `${TERM_POSTING_TAG}${keySegment(scope)}${keySegment(term)}`
}

/** Index partition key of the postings of `identifier` within `scope`. */
export function identifierPostingPk(scope: string, identifier: string): string {
  return `${IDENTIFIER_POSTING_TAG}${keySegment(scope)}${keySegment(identifier)}`
}

/** Index partition key holding one manifest per document of `scope`. */
export function manifestPk(scope: string): string {
  return `${MANIFEST_TAG}${keySegment(scope)}`
}

/**
 * Distinct posting partition keys of a document's terms followed by its identifiers, keeping the first
 * occurrence of each: a manifest written outside this library may repeat a term, and one transaction
 * cannot touch the same posting twice.
 */
export function postingPartitionKeys(
  scope: string,
  terms: readonly string[],
  identifiers: readonly string[]
): string[] {
  const termKeys = terms.map((term) => termPostingPk(scope, term))
  const identifierKeys = identifiers.map((identifier) => identifierPostingPk(scope, identifier))
  return [...new Set([...termKeys, ...identifierKeys])]
}

/**
 * Rejects a lone UTF-16 surrogate. It has no UTF-8 encoding, so its byte size and index keys would
 * silently differ from the Python package (which cannot encode it at all).
 */
function assertWellFormed(value: string, subject: string): void {
  if (LONE_SURROGATE.test(value)) {
    throw new StorageError(`${subject} must be well-formed Unicode; it contains a lone surrogate`)
  }
}

function wordRuns(text: string): string[] {
  const runs: string[] = []
  let current = ''
  for (const character of text) {
    if (isWordCharacter(character)) {
      current += character
    } else if (current) {
      runs.push(current)
      current = ''
    }
  }
  if (current) runs.push(current)
  return runs
}

function isWordCharacter(character: string): boolean {
  return JOINERS.has(character) || WORD_CHARACTER.test(character)
}

function runTerms(run: string): string[] {
  const compound = lowercaseAscii(run.replace(EDGE_JOINERS, ''))
  if (!compound) return []
  if (!JOINER.test(compound)) return [compound]
  return [compound, ...compound.split(JOINER_RUNS).filter(Boolean)]
}

function lowercaseAscii(value: string): string {
  return value.replace(ASCII_UPPERCASE_RUNS, (letters) => letters.toLowerCase())
}

File: python/src/strands_dynamodb_storage/lexical_terms.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

"""Pure ``lexical-v1`` rules: tokenizer, identifier normalization, index key encoding.

No I/O. Shared with the TypeScript package through ``testdata/lexical-v1.json`` so both
languages emit identical terms and index keys. Python ``str`` ordering is code-point order,
which equals UTF-8 byte order, so plain string sorting already matches DynamoDB sort keys.

Character classes come from the runtime's Unicode character database: code points assigned
after Unicode 15.0 may tokenize differently between runtimes (the fixtures pin assigned
characters only).
"""

from __future__ import annotations

import itertools
import re
import string
import unicodedata
from collections.abc import Iterable, Iterator, Sequence

from strands.types.exceptions import StorageError

LEXICAL_TOKENIZER_VERSION = "lexical-v1"

_JOINERS = "-_"
_JOINER_RUNS = re.compile(r"[-_]+")
_WORD_CATEGORY_CLASSES = frozenset("LMN")
_ASCII_LOWERCASE = str.maketrans(string.ascii_uppercase, string.ascii_lowercase)
_FIRST_SURROGATE = "\ud800"
_LAST_SURROGATE = "\udfff"

_TERM_POSTING_TAG = "t"
_IDENTIFIER_POSTING_TAG = "i"
_MANIFEST_TAG = "m"


def utf8_byte_length(value: str) -> int:
    """Return the UTF-8 encoded size of ``value`` in bytes.

    Never raises: a lone surrogate (not well-formed, so rejected by :func:`text_terms` and
    :func:`normalize_identifier`) counts 3 bytes, as the TypeScript package's ``TextEncoder``
    counts its U+FFFD replacement, so size checks behave identically in both languages.
    """
    return len(value.encode("utf-8", "surrogatepass"))


def text_terms(text: str) -> list[str]:
    """Split ``text`` into ``lexical-v1`` terms, de-duplicated in emission order.

    Runs of letters, marks, digits and the ASCII joiners ``-``/``_`` form words; everything
    else separates. Each word is emitted as a compound (joiners trimmed from the ends, ASCII
    letters lowercased, non-ASCII case kept) followed by its joiner-separated parts.

    Raises:
        StorageError: If the text is not well-formed Unicode (it contains a lone surrogate).
    """
    _require_well_formed(text, "Text")
    normalized = unicodedata.normalize("NFC", text)
    terms = itertools.chain.from_iterable(_word_terms(word) for word in _words(normalized))
    return list(dict.fromkeys(terms))


def normalize_identifier(identifier: str) -> str:
    """Return the NFC form of an exact-match identifier, otherwise unchanged.

    Identifiers stay case-sensitive and keep punctuation, leading zeroes and internal spaces.

    Raises:
        StorageError: If the identifier is not well-formed Unicode (it contains a lone surrogate),
            is empty, contains a control character, or starts or ends with a separator (whitespace).
    """
    _require_well_formed(identifier, "Identifier")
    normalized = unicodedata.normalize("NFC", identifier)
    if not normalized:
        raise StorageError("Identifier must not be empty")
    if any(unicodedata.category(char) == "Cc" for char in normalized):
        raise StorageError("Identifier must not contain control characters")
    if _is_separator(normalized[0]) or _is_separator(normalized[-1]):
        raise StorageError("Identifier must not start or end with whitespace")
    return normalized


def normalize_identifiers(identifiers: Iterable[str]) -> list[str]:
    """Normalize each identifier and de-duplicate, keeping the first occurrence.

    Raises:
        StorageError: If any identifier is invalid (see :func:`normalize_identifier`).
    """
    return list(dict.fromkeys(normalize_identifier(identifier) for identifier in identifiers))


def scope_segment(scope: str) -> str:
    """Length-prefixed form of ``scope`` (``"9:tenant/a/"``; ``"0:"`` for the root scope).

    Stored as the base item's scope marker so the root scope is never an empty string in a
    condition expression value.
    """
    return _segment(scope)


def term_posting_pk(scope: str, term: str) -> str:
    """Partition key of the posting list for ``term`` within ``scope``."""
    return f"{_TERM_POSTING_TAG}|{_segment(scope)}{_segment(term)}"


def identifier_posting_pk(scope: str, identifier: str) -> str:
    """Partition key of the posting list for an exact ``identifier`` within ``scope``."""
    return f"{_IDENTIFIER_POSTING_TAG}|{_segment(scope)}{_segment(identifier)}"


def manifest_pk(scope: str) -> str:
    """Partition key holding one manifest per indexed document of ``scope``."""
    return f"{_MANIFEST_TAG}|{_segment(scope)}"


def posting_partition_keys(scope: str, terms: Sequence[str], identifiers: Sequence[str]) -> list[str]:
    """Partition keys of every posting owned by a document with these terms and identifiers.

    Terms come first, then identifiers, de-duplicated keeping the first occurrence: a
    hand-edited manifest may repeat a term, and one transaction cannot touch an item twice.
    """
    partitions = itertools.chain(
        (term_posting_pk(scope, term) for term in terms),
        (identifier_posting_pk(scope, identifier) for identifier in identifiers),
    )
    return list(dict.fromkeys(partitions))


def _require_well_formed(value: str, subject: str) -> None:
    """Reject a lone surrogate (U+D800..U+DFFF): it has no UTF-8 encoding, so no stable term or key.

    The message names ``subject`` only, never the value.
    """
    if any(_FIRST_SURROGATE <= char <= _LAST_SURROGATE for char in value):
        raise StorageError(f"{subject} must be well-formed Unicode; it contains a lone surrogate")


def _segment(value: str) -> str:
    """Length-prefix ``value`` so concatenated segments can never collide."""
    return f"{utf8_byte_length(value)}:{value}"


def _is_word_char(char: str) -> bool:
    return char in _JOINERS or unicodedata.category(char)[0] in _WORD_CATEGORY_CLASSES


def _is_separator(char: str) -> bool:
    return unicodedata.category(char).startswith("Z")


def _words(text: str) -> Iterator[str]:
    for is_word, chars in itertools.groupby(text, key=_is_word_char):
        if is_word:
            yield "".join(chars)


def _word_terms(word: str) -> list[str]:
    """Compound term of ``word`` plus, when it contains joiners, each non-empty part."""
    compound = word.strip(_JOINERS).translate(_ASCII_LOWERCASE)
    if not compound:
        return []
    parts = _JOINER_RUNS.split(compound)
    return [compound, *parts] if len(parts) > 1 else [compound]
